components/scream/src/share/util/units.hpp

```diff
@@ -0,0 +1,287 @@
+}
+
+constexpr Units sqrt(const Units& x) {
+  return Units(x.m_units[0] / 2,
+               x.m_units[1] / 2,
+               x.m_units[2] / 2,
+               x.m_units[3] / 2,
+               x.m_units[4] / 2,
+               x.m_units[5] / 2,
+               x.m_units[6] / 2,
+               pow(x.m_scaling,RationalConstant(1,2)));
+}
+
+inline std::string to_string(const Units& x) {
+  std::string s;
+  int num_non_trivial = 0;
+  for (int i=0; i<NUM_BASIC_UNITS; ++i) {
+    if (x.m_units[i].num()==0) {
+      continue;
+    }
+    ++num_non_trivial;
+    s += " ";
+    s += BASIC_UNITS_SYMBOLS[i];
+    if (x.m_units[i]!=RationalConstant::one()) {
+      s += "^" + to_string(x.m_units[i],x.m_exp_format);
+    }
+  }
+
+  // Prepend the scaling only if not one or dimensionless unit
+  if (x.m_scaling!=RationalConstant::one() || num_non_trivial==0) {
+    s = to_string(x.m_scaling,x.m_exp_format) + s;
+  }
+  return s;
+}
+
+inline std::ostream& operator<< (std::ostream& out, const Units& x) {
+  out << to_string(x);
+  return out;
+}
+
+// ================== SHORT NAMES FOR COMMON PHYSICAL UNITS =============== //
+
+// Note to developers:
+// I added the 'most common' units. I avoided 'Siemes', since
+// the symbol is 'S', which is way too close to 's' (seconds).
+// TODO: should we add 'common' scaled ones, such as km=kilo*m?
+
+// === FUNDAMENTAL === //
+
+constexpr Units m   = Units(1,0,0,0,0,0,0);
+constexpr Units s   = Units(0,1,0,0,0,0,0);
+constexpr Units kg  = Units(0,0,1,0,0,0,0);
+constexpr Units K   = Units(0,0,0,1,0,0,0);
+constexpr Units A   = Units(0,0,0,0,1,0,0);
+constexpr Units mol = Units(0,0,0,0,0,1,0);
+constexpr Units cd  = Units(0,0,0,0,0,0,1);
+
+// === DERIVED === //
+
+// Thermomechanics
+constexpr auto day  = 86400*s;      // day          (time)
+constexpr auto year = 365*day;      // year         (time)
+constexpr auto g    = milli*kg;     // gram         (mass)
+constexpr auto N    = kg*m/(s*s);   // newton       (force)
+constexpr auto dyn  = N/(10000);    // dyne         (force)
+constexpr auto Pa   = N/(m*m);      // pascal       (pressure)
+constexpr auto bar  = 10000*Pa;     // bar          (pressure)
+constexpr auto atm  = 101325*Pa;    // atmosphere   (pressure)
+constexpr auto J    = N*m;          // joule        (energy)
+constexpr auto W    = J/s;          // watt         (power)
+
+// Electro-magnetism
+constexpr auto C    = A*s;          // coulomb      (charge)
+constexpr auto V    = J/C;          // volt         (voltage)
+constexpr auto T    = N/(A*m);      // tesla        (magnetic field)
+constexpr auto F    = C/V;          // farad        (capacitance)
+constexpr auto Wb   = V*s;          // weber        (magnetic flux)
+constexpr auto H    = Wb/A;         // henri        (inductance)
+constexpr auto Sv   = J/kg;         // sievert      (radiation dose)
+constexpr auto rem  = Sv/100;       // rem          (radiation dose)
+constexpr auto Hz   = 1/s;          // hertz        (frequency)
```

Comment:
How would we deal with mixing ratios in this framework?  Kg/Kg

Would the logic understand that Kg/Kg would be equivalent to Kg^0, ie unitless, or would the repeat of the string "Kg" break the logic?

Reply:
The units that cancel out do not appear in the string. The `to_string` method skips units that have a 0 exponent.

And btw, the check whether units are correct or not is done at the `Units` object level. The `to_string` functionality is just for output purposes. Granted, one can compare `to_string(u1)==to_string(u2)`, but the result _should_ be the same as `u1==u2`.

Reply:
Re: how to deal with mixing ratio. The framework would not be able to distinguish a mixing ration from any other dimensionless constant. Which is why I did not include `rad` (radian) and `sr` (steradian), since they are dimensionless units.

For mixing ratios, we can still get correctness by specifying numerator and denominator units, since we can catch different scales. A mixing ratio of `kg/kg` is the same as a mixing ratio of `g/g`. But if someone tries to set the units to `kg/g`, then that would be recognized as a different unit, due to the different scaling coefficient.